### ISSUE TITLE
Bug: @typescript-eslint/no-unnecessary-type-assertion False positive on non-null assertion after an implicitly-any variable gets initialised inside a conditional or loop

### ISSUE THREAD

**Issue by ExplodingCabbage** [typescript-eslint/typescript-eslint#11054]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.ts&code=GYVwdgxgLglg9mABAUwB4EMC2AHANsgCgEpEBvAKEUXykVQG5LEZhECBZdKACwDoAndGAAmcTMUQA%2BRAFoAjCQpUqqRAF5EAJkZUAvk37IoIfklQBCRAGpEcxvvJoseQkUZA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2uCUADgDYCuA5gJZKYAC8AnodGgMYBOlh8AtC8dXgB6dtFYB7ALaTkKaCh6NmPVgAsxAa3mR8AXQA0%2BSO1LEWWMMAC%2BIK0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
function example() {
  let x;
  if (Math.random() > -1) {
    x = 2;
  }
  return x! + 1;
}

example();
```

### ESLint Config

```javascript
import tseslint from 'typescript-eslint';

export default tseslint.config(
  tseslint.configs.recommendedTypeChecked,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.dirname,
      },
    },
  },
  {
    files: ['eslint.config.mjs'],
    extends: [tseslint.configs.disableTypeChecked],
  }
);
```

### tsconfig

```jsonc
{
    "include": ["**/*.ts"],
    "compilerOptions": {
        "strict": true,
        
    }
}
```

### Expected Result

No errors, since the non-null assertion (`!`) *is* necessary here; without it, tsc will refuse to compile the code, complaining that `'x' is possibly 'undefined'!

### Actual Result

```
6:10  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
```

### Additional Info

Things that seem necessary to produce the error above:

* `x` is not explicitly typed when declared (so is implicitly `any`)
* the initial assignment of `x` takes place inside a conditional or a loop, and TypeScript therefore can't tell, without the non-null assertion, that the initial assignment is guaranteed to take place
* `strict: true` is enabled

The example code above is of course contrived, but a more realistic example where this would in fact be totally reasonable code would be one where, instead of an `if (...) {...}` that's guaranteed to run, the assignment takes place within a *loop* that's guaranteed to run (and `x` gets used to store some sort of information about the final iteration of the loop that gets used after the loop finishes).

**Comment by ronami** [comment]
Interesting, seems like a missed edge case in [`isPossiblyUsedBeforeAssigned`](https://github.com/typescript-eslint/typescript-eslint/blob/23c5aa725ede26fbb88ac77cb950181ac2fca4d7/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L68-L149).
